kcqrs-core/src/main/kotlin/com/clouway/kcqrs/core/AggregateRootBase.kt

```diff
@@ -10,80 +11,126 @@ import java.util.*
  *
  * @author Miroslav Genov (miroslav.genov@clouway.com)
  */
-abstract class AggregateRootBase private constructor(@JvmField protected var aggregateId: String?) : AggregateRoot {
+abstract class AggregateRootBase private constructor(@JvmField protected var aggregateId: String? = "") : AggregateRoot {
 
-  private var changes: ArrayList<Event> = ArrayList()
-  private var version = 0L
+    private var changes: ArrayList<Event> = ArrayList()
+    private var version = 0L
 
-  constructor() : this(null)
+    constructor() : this(null)
 
-  override fun getId(): String? {
-    return aggregateId
-  }
+    override fun getId(): String? {
+        return aggregateId
+    }
 
-  override fun markChangesAsCommitted() {
-    changes.clear()
-  }
+    override fun <T : AggregateRoot> fromSnapshot(snapshotData: String): T {
+        val fromSnapshot = getSnapshotMapper().fromSnapshot(snapshotData)
+        val newInstance = this@AggregateRootBase::class.java.newInstance()
+        setFields(fromSnapshot, newInstance)
+        return newInstance as T
+    }
 
-  override fun getExpectedVersion(): Long {
-    return version
-  }
+    override fun getSnapshotMapper(): SnapshotMapper<AggregateRoot> {
+        //TODO(V.Mitov) Pass a serializer so that type adapters could be used
+        return object : SnapshotMapper<AggregateRoot> {
+            val gson = Gson()
+            override fun toSnapshot(data: AggregateRoot): String {
+                return gson.toJson(data)
+            }
+
+            override fun fromSnapshot(snapshot: String): AggregateRoot {
```

Comment:
Version of the snapshot could be passed as second parameter, so the clients will be able to distinguish between versions: 

Snapshot(payload, 1) 
Snapshot(payload, 2) 

```kotlin
if (version == 1)  {
   // read version 1 of the snapshot 
} else if (version == 2) {
  // read version 2 of the snapshot
}
```

Now by passing the payload directly, the user has to try to deserialize multiple times.

Reply:
Could be done, but how will this improve the deserialization. As i see it, the snapshot mapper will be the one responsible to handle the missing fields from previous snapshots.

Reply:
If a change to the aggregate was done, the user making the change will have to implement a snapshot mapper, different from the default, which will handle the missing fields. When a new change is made a change to the snapshot mapper will have to be done. This is a workflow that we will have to start using.

Reply:
It will make life easier for the snapshot mapper when there are more then 2 versions of the snapshot. 

1. Snapshot (1)
2. Snapshot (1)
3. Snapshot (2)
4. Snapshot (2)
5. Snapshot (3)
6. Snapshot (3)

e.g if changes between version 1 and 3 are major then the mapper will be able to check before the serialization what kind of object to provide for de-serialization. 

```kotlin
if (version == 1) {
   return gson.fromJson(OrderSnapshotV1::class.java).adapt()
} else if (version == 2) {
  return gson.fromJson(OrderSnapshotV2::class.java).adapt()
} else {
  return gson.fromJson(OrderSnapshotCurrent::class.java).adapt()
}
```

e.g the client code will be able to use different types instead of nullable and optional fields.

Reply:
Added the snapshot version to both the AggregateRoot and the SnapshotMapper.

Reply:
The AggregateRoot doesn't have to know about versions.